src/encoded/static/components/genetic_modification.js

```diff
@@ -316,190 +316,229 @@ Attribution.propTypes = {
+            modificationDocuments: [], // GM document search results
+            characterizationDocuments: [], // GM characterization document search results
+        };
+        this.requestDocuments = this.requestDocuments.bind(this);
+        this.geneticModificationComponentChanged = this.geneticModificationComponentChanged.bind(this);
+    }
+
+    componentDidMount() {
+        this.requestDocuments();
+    }
+
+    componentDidUpdate() {
+        this.requestDocuments();
+    }
+
+    requestDocuments() {
+        const { context } = this.props;
+
+        // Collect all GM characterization document @ids.
+        const characterizationDocumentAtIds = [];
+        if (context.characterizations && context.characterizations.length) {
+            context.characterizations.forEach((characterization) => {
+                if (characterization.documents && characterization.documents.length) {
+                    characterizationDocumentAtIds.push(...characterization.documents);
+                }
+            });
+        }
+
+        // Collect GM document @ids and combine with characterization document @ids so we can do
+        // one GET request for both.
+        const modificationCharacterizationDocumentsAtIds = _.uniq((context.documents && context.documents.length ? context.documents : []).concat(characterizationDocumentAtIds));
+
+        // Convert the array of document @ids into a query string and do the GET request for their
+        // objects.
+        const modificationCharacterizationDocumentsQuery = modificationCharacterizationDocumentsAtIds.length ? modificationCharacterizationDocumentsAtIds.reduce((acc, document) => `${acc}&${globals.encodedURIComponent(`@id=${document}`)}`, '') : null;
+        requestSearch(`type=Document${modificationCharacterizationDocumentsQuery}`).then(
+            (results) => {
+                // `results` is the search results object, or {} if it 404ed.
+                const modificationDocuments = [];
+                const characterizationDocuments = [];
+                if (results['@graph'] && results['@graph'].length) {
+                    // Split the results into arrays of modification and characterization
+                    // document objects.
+                    results['@graph'].forEach((document) => {
+                        if (characterizationDocumentAtIds.indexOf(document['@id']) > -1) {
+                            characterizationDocuments.push(document);
+                        } else {
+                            modificationDocuments.push(document);
+                        }
+                    });
+                }
+
+                // If the resulting documents are different from what we have in state, then
+                // update the state.
+                if (this.geneticModificationComponentChanged(modificationDocuments, characterizationDocuments)) {
+                    this.setState({ modificationDocuments, characterizationDocuments });
```

Comment:
It feels odd to do it this way when React already has `shouldComponentUpdate` that is called when the state changes.

Reply:
Was wondering whether to add a comment on that, and I guess I should have. The timing of shouldComponentUpdate isn’t quite right for this case, and solving that would have been, in my view, much less clear than just saying, “If the results are the same as what we had in state, then that’s it for re-rendering until something else changes.” I checked other resources to see if there was some clever way, but I mostly saw people just adding an if condition on the setState instead of using shouldComponentUpdate. It’s great when props are changing, but in this case we don’t know what’s changing until the GET request comes back.